### PR TITLE
[SPARK-45232][SQL][DOCS] Add missing function groups to SQL references

### PR DESCRIPTION
**Pull request by zhengruifeng**
### What changes were proposed in this pull request?
Add missing function groups to SQL references:
- xml_funcs
- lambda_funcs
- collection_funcs
- url_funcs
- hash_funcs
- struct_funcs

Note that this PR doesn't fix `table_funcs`:
1, `gen-sql-functions-docs.py` doesn't work properly with `TableFunctionRegistry`, I took a cursory look but fail to fix it;
2, table functions except `range` (e.g. `explode`) were already contained in `Generator Functions`, not sure we need to show them twice.

### Why are the changes needed?
when referring to the SQL references, I find many functions are missing https://spark.apache.org/docs/latest/sql-ref-functions.html.

### Does this PR introduce _any_ user-facing change?
yes

### How was this patch tested?
manually check

### Was this patch authored or co-authored using generative AI tooling?
no

### CODE REVIEW

File: sql/gen-sql-functions-docs.py [hunk 1]
```diff
@@ -34,6 +34,8 @@
     "math_funcs", "conditional_funcs", "generator_funcs",
     "predicate_funcs", "string_funcs", "misc_funcs",
     "bitwise_funcs", "conversion_funcs", "csv_funcs",
+    "xml_funcs", "lambda_funcs", "collection_funcs",
+    "url_funcs", "hash_funcs",
```
Comment (zhengruifeng) [inline sql/gen-sql-functions-docs.py:38]: manually check with
```
ag --scala 'group = \"' sql
```

all groups but `table_funcs` should be here

File: sql/gen-sql-functions-docs.py [hunk 2]
```diff
@@ -34,6 +34,8 @@
     "math_funcs", "conditional_funcs", "generator_funcs",
     "predicate_funcs", "string_funcs", "misc_funcs",
     "bitwise_funcs", "conversion_funcs", "csv_funcs",
+    "xml_funcs", "lambda_funcs", "collection_funcs",
+    "url_funcs", "hash_funcs", "struct_funcs",
```
Comment (zhengruifeng) [inline sql/gen-sql-functions-docs.py:38]: check against
https://github.com/apache/spark/blob/37ab190dc5bfa59b4e06af9551c35ab179a05733/sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/ExpressionInfo.java#L43-L48

two difference:
1, `table_funcs`: not support in `gen-sql-functions-docs.py`;
2, `binary_funcs`: we don't have function using this group;

**Review by HyukjinKwon — APPROVED** [review]

File: sql/gen-sql-functions-docs.py (hunk 2 above)
Comment (allisonwang-db) [reply sql/gen-sql-functions-docs.py:38]: QQ: For generator_funcs, do we have documentation for them when used in the FROM clause of a query? Functions like explode are typically considered table-valued generator functions.

**Review by allisonwang-db — APPROVED** [review]

**Comment by zhengruifeng** [comment]
@allisonwang-db I am not sure, I don't see document for FROM clause, you may check 3 places:

- https://spark.apache.org/docs/latest/api/sql/index.html#explode
- https://spark.apache.org/docs/latest/sql-ref-functions-builtin.html#generator-functions
- https://spark.apache.org/docs/latest/sql-ref-syntax.html

**Comment by gatorsmile** [comment]
cc @srielau

**Comment by zhengruifeng** [comment]
image

image

image

image

image

image

image

**Comment by zhengruifeng** [comment]
we can check the documents built in the GA of this PR,  https://github.com/zhengruifeng/spark/actions/runs/6249096629

image


however, it expires after one day

**Comment by srielau** [comment]
What is the purpose of "lambda function"? All others are type-specific or "functionality"-specific.
But lambda is "technology". What is the user journey that would drive one to browse lambda functions?

**Comment by zhengruifeng** [comment]
> What is the purpose of "lambda function"? All others are type-specific or "functionality"-specific. But lambda is "technology".

lambda functions were already exposed to end users (e.g. https://spark.apache.org/docs/latest/api/python/reference/pyspark.sql/api/pyspark.sql.functions.array_sort.html#pyspark.sql.functions.array_sort). 

I think if we document other functions here, it is better to add lambda functions as well.

> What is the user journey that would drive one to browse lambda functions?

I think this could be an example: when a user try to sort array of structs by a specific order, he may refer to the document of `array_sort`

**Comment by HyukjinKwon** [comment]
Just to be clear, this is automatic documentation based on the current documentation. If the grouping is wrong, or to be fixed, we should fix `functions.scala` main documentation.

**Comment by zhengruifeng** [comment]
@HyukjinKwon this page is not built from `functions.scala`, but from the groups specified in expression definitions, like

https://github.com/apache/spark/blob/6b747ab8cef44ea46102b2685b9254f7c4828ea8/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/mathExpressions.scala#L1313

**Comment by HyukjinKwon** [comment]
yeah, I mean individual `ExpressionDescription`.

**Comment by srielau** [comment]
> > What is the purpose of "lambda function"? All others are type-specific or "functionality"-specific. But lambda is "technology".
> 
> lambda functions were already exposed to end users (e.g. https://spark.apache.org/docs/latest/api/python/reference/pyspark.sql/api/pyspark.sql.functions.array_sort.html#pyspark.sql.functions.array_sort).
> 
> I think if we document other functions here, it is better to add lambda functions as well.
> 
> > What is the user journey that would drive one to browse lambda functions?
> 
> I think this could be an example: when a user try to sort array of structs by a specific order, he may refer to the document of `array_sort`

If I try to find a function that sorts arrays I will try to find that function under collection functions.
Just like substr() is not a math function, even though most of its arguments are integers. Substr operates on strings...
array_sort operates on arrays.

**Comment by zhengruifeng** [comment]
> > > What is the purpose of "lambda function"? All others are type-specific or "functionality"-specific. But lambda is "technology".
> > 
> > 
> > lambda functions were already exposed to end users (e.g. https://spark.apache.org/docs/latest/api/python/reference/pyspark.sql/api/pyspark.sql.functions.array_sort.html#pyspark.sql.functions.array_sort).
> > I think if we document other functions here, it is better to add lambda functions as well.
> > > What is the user journey that would drive one to browse lambda functions?
> > 
> > 
> > I think this could be an example: when a user try to sort array of structs by a specific order, he may refer to the document of `array_sort`
> 
> If I try to find a function that sorts arrays I will try to find that function under collection functions. Just like substr() is not a math function, even though most of its arguments are integers. Substr operates on strings... array_sort operates on arrays.

Got it.
I think we can rename the group name in `.md` files, but need to be different from others. What about `Advanced Collection Functions`? @srielau

**Comment by srielau** [comment]
> > > > What is the purpose of "lambda function"? All others are type-specific or "functionality"-specific. But lambda is "technology".
> > > 
> > > 
> > > lambda functions were already exposed to end users (e.g. https://spark.apache.org/docs/latest/api/python/reference/pyspark.sql/api/pyspark.sql.functions.array_sort.html#pyspark.sql.functions.array_sort).
> > > I think if we document other functions here, it is better to add lambda functions as well.
> > > > What is the user journey that would drive one to browse lambda functions?
> > > 
> > > 
> > > I think this could be an example: when a user try to sort array of structs by a specific order, he may refer to the document of `array_sort`
> > 
> > 
> > If I try to find a function that sorts arrays I will try to find that function under collection functions. Just like substr() is not a math function, even though most of its arguments are integers. Substr operates on strings... array_sort operates on arrays.
> 
> Got it. I think we can rename the group name in `.md` files, but need to be different from others. What about `Advanced Collection Functions`? @srielau

How about having our cake and eat it to? Can a function be in more than one group?

**Comment by zhengruifeng** [comment]
> > > > > What is the purpose of "lambda function"? All others are type-specific or "functionality"-specific. But lambda is "technology".
> > > > 
> > > > 
> > > > lambda functions were already exposed to end users (e.g. https://spark.apache.org/docs/latest/api/python/reference/pyspark.sql/api/pyspark.sql.functions.array_sort.html#pyspark.sql.functions.array_sort).
> > > > I think if we document other functions here, it is better to add lambda functions as well.
> > > > > What is the user journey that would drive one to browse lambda functions?
> > > > 
> > > > 
> > > > I think this could be an example: when a user try to sort array of structs by a specific order, he may refer to the document of `array_sort`
> > > 
> > > 
> > > If I try to find a function that sorts arrays I will try to find that function under collection functions. Just like substr() is not a math function, even though most of its arguments are integers. Substr operates on strings... array_sort operates on arrays.
> > 
> > 
> > Got it. I think we can rename the group name in `.md` files, but need to be different from others. What about `Advanced Collection Functions`? @srielau
> 
> How about having our cake and eat it to? Can a function be in more than one group?

probably we can. I will try to map `lambda` to `collection` just in the doc build.

I think making a function in more than one group would be much complex.

**Comment by zhengruifeng** [comment]
image


image


now we put `lambda functions` into `collection functions`

**Comment by zhengruifeng** [comment]
@srielau I have put `lambda functions` in `collection functions`, I think this PR is ready to merge?

**Comment by zhengruifeng** [comment]
thanks @srielau @allisonwang-db @HyukjinKwon 

merged to master

**Pull request closed without merging**